### PR TITLE
utility scripts

### PR DESCRIPTION
**Pull request by tnixa**
Added some utility scripts.

DCO 1.1 Signed-off-by: Terrence Nixa [tnixa@us.ibm.com](mailto:tnixa@us.ibm.com)

### CODE REVIEW

File: deleteAllWords.py [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+word_list = []
+words_raw = text_to_speech.list_words(customization_id).get_result()
+for element in words_raw['words']:
+    word_list.append(element['word'])
+print("existing word list: ", word_list)
+
+# loop through deleting the words
+for word in word_list:
+    text_to_speech.delete_word(customization_id, word)
+    print("deleted word: ", word)
```
Comment (andrewrfreed) [inline deleteAllWords.py:37]: Let's chat, either we always print progress or never, in these tools.

File: problematic_word_extraction.py [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+            if word not in word_list:
+                word_list.append(word)
+
+print("Length of word list is: ", len(word_list) )
+
+dash_list = []
+slash_list =[]
+pound_list = []
+colon_list = []
+parnthesis_list = []
```
Comment (andrewrfreed) [inline problematic_word_extraction.py:51]: spelling

File: problematic_word_extraction.py [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+# Divide the words up into separate files by punctuation
+
+import pandas as pd
+# may need to do a 'pip install pyspellchecker'
+from spellchecker import SpellChecker 
+import sys
+
+# python problematic_word_extraction.py messages.csv
+
+spell = SpellChecker()
```
Comment (andrewrfreed) [inline problematic_word_extraction.py:11]: Is this locale-dependent?

Comment (andrewrfreed) [reply problematic_word_extraction.py:11]: ie, does it work outside of US-English?

File: createCustomModel.py [hunk 4]
```diff
@@ -0,0 +1,37 @@
+import sys
+import json
+from config import Config
+from ibm_watson import TextToSpeechV1
+from ibm_cloud_sdk_core.authenticators import IAMAuthenticator
+
+# python createCustomModel.py config.ini "script test" en-US "test the script for creating a new model"
+
+config_file = "config.ini"
```
Comment (andrewrfreed) [inline createCustomModel.py:9]: For a few of these model scripts, we can consolidate to one like we did in STT-WER-Python-Tool: https://github.com/IBM/watson-stt-wer-python/blob/main/models.py

File: add_acronym_pronounce_to_json.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
+# Use csv of acronyms and translate, e.g. IBM -> I B M,
```
Comment (andrewrfreed) [inline add_acronym_pronounce_to_json.py:1]: Let's consolidate to camelCase or snake_case for the file names.

**Review by andrewrfreed — CHANGES_REQUESTED** [review]
These are really cool tools, I am anxious to use these.  Please take a look at the other review comments which will help normalize the utilities.

After that, I want to make sure it is still obvious how to find the "main" python scripts in this folder (the top ~3)

File: problematic_word_extraction.py (hunk 2 above)
Comment (tnixa) [reply problematic_word_extraction.py:51]: fixed

File: problematic_word_extraction.py (hunk 3 above)
Comment (tnixa) [reply problematic_word_extraction.py:11]: different languages are supported...
https://pyspellchecker.readthedocs.io/en/latest/
> The currently supported dictionaries are:

English - ‘en’
Spanish - ‘es’
French - ‘fr’
Portuguese - ‘pt’
German - ‘de’
Russian - ‘ru’

File: add_acronym_pronounce_to_json.py (hunk 5 above)
Comment (tnixa) [reply add_acronym_pronounce_to_json.py:1]: changed to snake_case

File: deleteAllWords.py (hunk 1 above)
Comment (tnixa) [reply deleteAllWords.py:37]: I prefer to print progress. I believe the "main" scripts print progress as well. This scripts specifically should be moved to "models.py" as mentioned in another comment

File: README.md [hunk 6]
```diff
@@ -161,3 +161,67 @@ extract_dialog = True
...
+Sample script execution: `python util_add_acronym_pronounce_to_json.py test_terms.csv CustomWords.json`
+
+### util_add_word_phonetic_to_json.py
+Parse phonetic pronunciations from the csv output file from `pronounce.py`  
+Then add the tranlations to an existing custom words TTS json file.  
+Sample script execution: `python util_add_word_phonetic_to_json.py tts_pronounce.csv CustomWords.json`
+
+### util_create_synthesis_input_file.py
+Create a synthesis input csv file from a csv file of words to synthesize.
+Sample script execution: `python util_createSynthesisInputFile.py words.csv inputfile_to_synthesize.csv `
```
Comment (andrewrfreed) [inline README.md:223]: Slight mismatch to actual python script name.

File: util_create_synthesis_input_file.py [hunk 7]
```diff
@@ -0,0 +1,18 @@
+import pandas as pd
+import sys
+
+# python createSynthesisInputFile.py words.csv inputfile_to_synthesize.csv
```
Comment (andrewrfreed) [inline util_create_synthesis_input_file.py:4]: File name is mismatched

File: util_problematic_word_extraction.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
+# Extract potentially problematic words for TTS from a CSV using a spell checker
+# Divide the words up into separate files by punctuation
+
+import pandas as pd
```
Comment (andrewrfreed) [inline util_problematic_word_extraction.py:4]: Need to add `pandas` and `pyspellchecker` to `requirements.txt`

File: util_problematic_word_extraction.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+# may need to do a 'pip install pyspellchecker'
+from spellchecker import SpellChecker 
+# https://pyspellchecker.readthedocs.io/en/latest/
+import sys
+
+# python problematic_word_extraction.py messages.csv
+
+spell = SpellChecker()
```
Comment (andrewrfreed) [inline util_problematic_word_extraction.py:12]: I would like to expose an additional parameter so we can support non-English languages:
https://pyspellchecker.readthedocs.io/en/latest/#non-english-dictionaries

File: README.md [hunk 10]
```diff
@@ -161,3 +161,67 @@ extract_dialog = True
...
+```
+
+Note some parameter combinations are not possible.  The operations supported all wrap the SDK methods documented at https://cloud.ibm.com/apidocs/text-to-speech.
+
+
+## Utility Scripts
+### util_add_acronym_pronounce_to_json.py
+Process a csv of acronyms and translate them, e.g. IBM -> I B M,  
+Then add the tranlations to an existing custom words TTS json file.  
+Sample script execution: `python util_add_acronym_pronounce_to_json.py test_terms.csv CustomWords.json`
```
Comment (andrewrfreed) [inline README.md:214]: For each script can you add a comment about its lifecycle?  Including how the inputs/outputs from one script feed into other repository scripts.

I would imagine you bookend this script with a `models.py get words` and `models.py update words` (with same JSON file in all three)

**Review by andrewrfreed — CHANGES_REQUESTED** [review]
Overall looks really good, I like the models.py updates especially.  A few cosmetic and doc changes and we are good to go.
Can you also remove/.gitignore the `.DS_Store` file?

File: README.md (hunk 6 above)
Comment (tnixa) [reply README.md:223]: fixed

File: util_create_synthesis_input_file.py (hunk 7 above)
Comment (tnixa) [reply util_create_synthesis_input_file.py:4]: fixed

File: util_problematic_word_extraction.py (hunk 8 above)
Comment (tnixa) [reply util_problematic_word_extraction.py:4]: added

File: README.md (hunk 6 above)
Comment (tnixa) [reply README.md:223]: fixed

File: util_problematic_word_extraction.py (hunk 9 above)
Comment (tnixa) [reply util_problematic_word_extraction.py:12]: added

File: README.md (hunk 10 above)
Comment (tnixa) [reply README.md]: I took a stab at this

File: deleteAllWords.py (hunk 1 above)
Comment (andrewrfreed) [reply deleteAllWords.py]: Moved print to stderr so that stdout can be safely redirected to output file.

File: createCustomModel.py (hunk 4 above)
Comment (andrewrfreed) [reply createCustomModel.py]: File is removed.

File: problematic_word_extraction.py (hunk 3 above)
Comment (andrewrfreed) [reply problematic_word_extraction.py]: Great, thanks for adding!

File: util_problematic_word_extraction.py (hunk 8 above)
Comment (andrewrfreed) [reply util_problematic_word_extraction.py:27]: I got error installing:

```
ERROR: Ignored the following versions that require a different python version: 1.4.0 Requires-Python >=3.8; 1.4.0rc0 Requires-Python >=3.8; 1.4.1 Requires-Python >=3.8; 1.4.2 Requires-Python >=3.8; 1.4.3 Requires-Python >=3.8; 1.5.0rc0 Requires-Python >=3.8
```

Seems like `pandas>=1.0.0` suffices so I downgraded the dependency.

File: README.md (hunk 10 above)
Comment (andrewrfreed) [reply README.md]: The readme sections give hints based on the parameters how they fit into the lifecycle.

**Review by andrewrfreed — APPROVED** [review]

**Comment by andrewrfreed** [comment]
I pushed several revisions so I will add my DCO also:

DCO 1.1 Signed-off-by: Andrew R. Freed [afreed@us.ibm.com](mailto:afreed@us.ibm.com)

**Comment by andrewrfreed** [comment]
Passing review back to you @tnixa for all the changes I made.  Comment back when you like it and we can merge.

**Review by andrewrfreed — APPROVED** [review]
Great job, thanks for all the revisions.  Let's merge!

**Pull request merged**